File: java/Cards.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.Comparator;

 class Cards implements Comparable<Cards>{
	
	private String rank;
	private char suit;
	
	//Constructor
	public Cards(String rank, char suit) {
		this.rank = rank;
		this.suit = suit;
		
			
	}
	

  public String getRank() {
			return rank;
  }
	
	
	public char getSuit() {
		return suit;
	}
 
 
	//overide the compareTo method to compare each hand
	@Override  
    public int compareTo(Cards o) {
        return this.rank.compareTo(o.rank);
    }
	
	//using an overridden toString method to return the card in order of the number, then rank
	@Override
	public String toString() {
		return rank +""+ suit;
		
	}
 
 //using switch cases to change the string value to the value it represents
 
 public static int getRankValue(String rank){
   switch(rank) {
       case "2": return 2;
       case "3": return 3;
       case "4": return 4;
       case "5": return 5;
       case "6": return 6;
       case "7": return 7;
       case "8": return 8;
       case "9": return 9;
       case "10": return 10;
       case "J": return 11;
       case "Q": return 12;
       case "K": return 13;
       case "A": return 14;
       default: return -1;
   }
 }
	


}

/////DECK CLASS///////////////////////////////////////////////////////////////////////////


	class Deck {
   //creating a private final variable since the deck will never change
   private final List<Cards> cards;
   
   public Deck() {
   cards = new ArrayList<>();
   char[] suits = { 'S' , 'H' , 'D', 'C'};
   String [] ranks = { "2", "3", "4", "5" , "6", "7", "8", "9", "10" , "J", "Q", "K", "A"};
   
   for (char suit: suits) {
     for (String rank: ranks){
       cards.add(new Cards(rank, suit));
       }
     }
   
   
   }
   
 //shuffles the deck
  public void shuffle(){
    Collections.shuffle(cards);
  
  } 
  
  //dealing cards
  public Cards dealingCards(){
    if (!cards.isEmpty())
    {
      return cards.remove(0);
    }
    return null;
  }
  
  //retruns the size of the deck
  public int size(){
    return cards.size();
  
  }
  
  // printing the deck
  public void printDeck(){
  int count = 0;
  
  for (Cards card: cards) {
    System.out.print(card.toString() + "\t");
    count++;
    if (count % 13 ==0) {
        System.out.println();
    }
  }
 }
 
 //printing what is remaining in the deck
 public void printRemainingDeck(){
   System.out.println();
   System.out.println("Remaining Cards in Deck: " + "\t"); 
   for (Cards card: cards) {
     System.out.print(card.toString() + "\t");
     }
   }
  
 }
 
 
   //HAND CLASS//////////////////////////////////////////////////////////////
   
 class Hand implements Comparable<Hand> {
   private final List<Cards> cards;
   private String handRank;
   
   public Hand() {
     cards = new ArrayList<>();
     Collections.sort(cards); 
     
   }
   
   //adding a card to the hand
   public void addCard (Cards card) {
     cards.add(card);
   
   }
   
   //returning the cards
   public List<Cards> getCards() {
     return cards;
     
   }
 
    public void sortHand() {
        Collections.sort(cards, new Comparator<Cards>() {
        
        @Override
        public int compare(Cards card1, Cards card2) {
            return card1.getSuit() - card2.getSuit();
          }
        });
      } 
      
    //overide the compareTo method to compare each hand  
    @Override  
    public int compareTo(Hand o) {
        return this.rankHand().compareTo(o.rankHand());
    }
       
   
   //printing out the hands
   public void printHand(){
       int count = 0;
       for (Cards newCard: cards){
           System.out.print(newCard.toString() + "\t");
           count++;
           if (count % 5 == 0){
               System.out.println();
           }
       }
   }
               
   
   @Override 
   public String toString() {
   // StringBuilder allows to store each card to make a hand and return it
   
     StringBuilder handString = new StringBuilder();
     int count = 0;
     for (Cards card : cards) {
       handString.append(card.toString()).append("\t");
       count++;
       if (count % 5 == 0){
         System.out.println();
       }
     }
   return handString.toString();
 }
 
   public String rankHand() {
     //Creating a hash map to rank them hands
     Map<Character, Integer> suitCounts = new HashMap<>(); //for the suits
     Map<String, Integer> rankCounts = new HashMap<>();  //for the ranks
     this.handRank = "Royal Straight Flush";
     for (Cards newCard: cards) {
       char suit = newCard.getSuit();
       String rank = newCard.getRank();
       suitCounts.put(suit, suitCounts.getOrDefault(suit, 0) + 1);
       rankCounts.put(rank, rankCounts.getOrDefault(rank, 0) + 1);
     }
     
     if (suitCounts.size() == 1 && rankCounts.size() == 5) {
         if (isRoyalStraightFlush(rankCounts.keySet(), suitCounts.keySet().iterator().next())) {
             return "Royal Straight Flush";
         }
         
         return "Straight Flush";
     }
     
     if (suitCounts.size() == 1 &&  rankCounts.size() == 5) {
         if (isStraight(rankCounts.keySet())) {
         return "Straight Flush";
       }
       return "Flush";
     }
     
     if (rankCounts.containsValue(4)) {
         return "Four of a Kind"; //4 of each number (A, 2, 3....)
     }
     
     if (rankCounts.containsValue(3) && rankCounts.containsValue(2)) {
         return "Full House";  //Full House, three of a kind and a pair
     }
     
     if (rankCounts.size() == 5 && isStraight(rankCounts.keySet())) {
       return "Straight" ; //5 in a row
     }
     
     
     if (rankCounts.containsValue(3)) {
         return "Three of a Kind"; // 3 of each number
     }
     
 
     if (rankCounts.containsValue(2)) { // for getting pairs
         int pairCount = 0;
         for (int newCount : rankCounts.values()) {
             if (newCount == 2) {
                 pairCount++;
             }
         }
         if (pairCount == 2) {
             return "Two Pair";
         } else {
             return "Pair"; //2 of a rank
         }
     }
     
     return "High Card"; //the highest card
     
     
     String rank = rankHand();
     
     if (rank.equals("Royal Straight Flush") || rank.equals("Straight Flush") || rank.equals("Flush")) {
         List<Cards> sortedCards = SCBR(cards);
         Cards highestCard = sortedCards.get(sortedCards.size() - 1);
         this.handRank += " - Highest Card Suit: " + highestCard.getSuit();
     } else if (rank.equals("Straight")) {
         List<Cards> sortedCards = SCBR(cards);
         Cards highestCard = sortedCards.get(sortedCards.size() - 1);
         this.handRank += " - Straight: Highest Card Suit: " + highestCard.getSuit();
     } else if (rank.equals("Two Pair")) {
         List<Cards> sortedCards = SCBR(cards);
         Cards highPairCard = null;
         Cards lowPairCard = null;
         Cards kickerCard = null;
         for (Cards card : sortedCards) {
             int frequency = Collections.frequency(sortedCards, card);
             if (frequency == 2) {
                 if (highPairCard == null) {
                      highPairCard = card;
                 } else if (lowPairCard == null) {
                      lowPairCard = card;
                 }
             } else if (frequency == 1) {
                  kickerCard = card;
             }
         }
         
         this.handRank += " -Two Pair: Kicker Suit: " + kickerCard.getSuit();
     } else if (rank.equals("Pair")) {
         List<Cards> sortedCards = SCBR(cards);
         Cards pairCard = null;
         Cards kickerCard = null;
         for (Cards card : sortedCards) {
             int frequency = Collections.frequency(sortedCards, card);
             if (frequency == 2) {
                 pairCard = card;
             } else if (frequency == 1) {
                 kickerCard = card;
             }
         }
         
         this.handRank += " - Pair: Kicker Suit: " + kickerCard.getSuit();
     } else if (rank.equals("High Card")) {
          List<Cards> sortedCards = SCBR(cards);
          Cards highestCard = sortedCards.get(sortedCards.size() - 1);
          this.handRank += " - High Card: Suit: " + highestCard.getSuit();
     }
         
              
   }
  //sortCardsByRank       
 private List<Cards> SCBR(List<Cards> cards) { 
     Collections.sort(cards, new Comparator<Cards>() {
         @Override
         public int compare(Cards card1, Cards card2) {
           return card1.getRank().compareTo(card2.getRank());
          }
        });
        return cards;
 }  
   
 public char getFlushSuit() {
     Map<Character, Integer> suitCounts = new HashMap<>();
     for (Cards newCard : cards) {
         char newSuit = newCard.getSuit();
         suitCounts.put(newSuit, suitCounts.getOrDefault(newSuit, 0) + 1);
     }
     
     for (char newSuit : suitCounts.keySet()) {
         if (suitCounts.get(newSuit) >= 5) {
             return newSuit;
         }
     }
     
     return ' ' ; // there is No flush 
 }
 
 
 

 
 
 private boolean isRoyalStraightFlush(Set<String> ranks, char suit) {
     //checking for the sequence of 10, J, Q, K, A
     
     return ranks.containsAll(Arrays.asList("10", "J", "Q", "K", "A"));
 }
 
 
 private boolean isStraight(Set<String> ranks) {
     List<Integer> rankValues = new ArrayList<>();
     for (String newRank : ranks) {
         rankValues.add(Cards.getRankValue(newRank));
     }
     
     Collections.sort(rankValues);
     
     int min = rankValues.get(0);
     int max = rankValues.get(rankValues.size() - 1);
     
     if (max - min == 4 && ranks.size() == 5) {
         return true;
     }
     
     //check for an Ace low straight (A, 2, 3, 4, 5)
     if (ranks.contains("A") && ranks.contains("2") && ranks.contains("3") && ranks.contains("4") && ranks.contains("5")) {
         return true;
     }
     
     
     return false;


   }
 }

   

   //PLAYER CLASS////////////////////////////////////////////////////////////////////
   
  class Player {
   private final String name;
   private final Hand hand;
   
 
   public Player (String name) {
     this.name = name;
     this.hand = new Hand();
     
   }
   
   public void addingToHand(Cards newCard)
   {
     hand.addCard(newCard);
   }
   
   public String getName() {
       return name;
   }
   
   public Hand getHand(){
     return hand;
   }
   
   @Override
   public String toString() {
     //return name + " Hand: \n" + hand.toString();
     return name;
     
   }
 }
 
 class PlayerRank implements Comparable<PlayerRank> {
   private final Player player;
   private final String rank;
   private final char suit; //used for tiebreakers
   
   public PlayerRank(Player player, String rank, char suit) {
         this.player = player;
         this.rank = rank;
         this.suit = suit;
   }
   
   public Player getPlayer() {
       return player;
   }
   
   public String getRank() {
       return rank;
   }
   
   @Override
    public int compareTo(PlayerRank other) {
        return this.rank.compareTo(other.rank);
    }
   
   public char getSuit() {
       return suit;
   }

   
   
 
 
 
 
 
 
 }
 
 

   
 
 
 
 
 
 
 
 
 
 
 
 

   
   
 

File: python/Cards.py
from random import shuffle

# Cards Class
class Cards:
    def __init__(self, rank, suit):
        self.rank = rank
        self.suit = suit

    def get_rank(self):
        return self.rank

    def get_suit(self):
        return self.suit

    def __lt__(self, other):
        return self.rank < other.rank

    def __str__(self):
        return str(self.rank) + str(self.suit)

    @staticmethod
    def get_rank_value(rank):
        rank_values = {
            "2": 2, "3": 3, "4": 4, "5": 5, "6": 6, "7": 7,
            "8": 8, "9": 9, "10": 10, "J": 11, "Q": 12, "K": 13, "A": 14
        }
        return rank_values.get(rank, -1)


# Deck Class
class Deck:
    def __init__(self):
        self.cards = []
        suits = ['S', 'H', 'D', 'C']
        ranks = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"]

        for suit in suits:
            for rank in ranks:
                self.cards.append(Cards(rank, suit))
# suffling the deck
    def shuffle(self):
        shuffle(self.cards)
# dealing each card from the deck
    def deal_card(self):
        if self.cards:
            return self.cards.pop(0)
        return None
# returning the size of the deck
    def size(self):
        return len(self.cards)
# printing the deck
    def print_deck(self):
        count = 0
        line = []
        for card in self.cards:
            print(card, "\t")
            count += 1
            if count % 13 == 0:
                print()
# after getting the hands, printing what is remianing of the deck
    def print_remaining_deck(self):
        print("\nRemaining Cards in Deck:\t", "")
        for card in self.cards:
            print(card, "\t")
        print()


# Hand Class
class Hand:
    def __init__(self):
        self.cards = []
# adding a card to each hand
    def add_card(self, card):
        self.cards.append(card)
# getting the cards for each hand
    def get_cards(self):
        return self.cards
# sorting the hand in order of suit
    def sort_hand(self):
        self.cards.sort(key=lambda card: card.get_suit())

    def __lt__(self, other):
        return self.rank_hand() < other.rank_hand()
# printing each hand
    def print_hand(self):
        count = 0
        for card in self.cards:
            print(card, "\t")
            count += 1
            if count % 5 == 0:
                print()

    def __str__(self):
        hand_string = ""
        count = 0
        for card in self.cards:
            hand_string += str(card) + "\t"
            count += 1
            if count % 5 == 0:
                hand_string += "\n"
        return hand_string
# getting the ranks according to the rules of 5 card stud
    def rank_hand(self):
        suit_counts = {}
        rank_counts = {}
        hand_rank = "Royal Straight Flush"

        for card in self.cards:
            suit = card.get_suit()
            rank = card.get_rank()
            suit_counts[suit] = suit_counts.get(suit, 0) + 1
            rank_counts[rank] = rank_counts.get(rank, 0) + 1

        if len(suit_counts) == 1 and len(rank_counts) == 5:
            if self.is_royal_straight_flush(rank_counts.keys(), next(iter(suit_counts))):
                return "Royal Straight Flush"
            return "Straight Flush"

        if len(suit_counts) == 1 and len(rank_counts) == 5:
            if self.is_straight(rank_counts.keys()):
                return "Straight Flush"
            return "Flush"

        if 4 in rank_counts.values():
            return "Four of a Kind"

        if 3 in rank_counts.values() and 2 in rank_counts.values():
            return "Full House"

        if len(rank_counts) == 5 and self.is_straight(rank_counts.keys()):
            return "Straight"

        if 3 in rank_counts.values():
            return "Three of a Kind"

        if 2 in rank_counts.values():
            pair_count = sum(1 for count in rank_counts.values() if count == 2)
            if pair_count == 2:
                return "Two Pair"
            else:
                return "Pair"

        return "High Card"

    def is_royal_straight_flush(self, ranks, suit):
        return all(rank in ranks for rank in ["10", "J", "Q", "K", "A"])

    def is_straight(self, ranks):
        rank_values = [Cards.get_rank_value(rank) for rank in ranks]
        rank_values.sort()
        min_rank = rank_values[0]
        max_rank = rank_values[-1]

        if max_rank - min_rank == 4 and len(ranks) == 5:
            return True

        if "A" in ranks and "2" in ranks and "3" in ranks and "4" in ranks and "5" in ranks:
            return True

        return False
        
        
    def get_flush_suit(self):
        suit_counts = {}
        for card in self.cards:
            suit = card.get_suit()
            if suit in suit_counts:
                suit_counts[suit] += 1
            else:
                suit_counts[suit] = 1

        for suit, count in suit_counts.items():
            if count >= 5:
                return suit

        return None  # No flush suit found

# Player Class
class Player:
    def __init__(self, name):
        self.name = name
        self.hand = Hand()
# adding a card to each player
    def add_to_hand(self, card):
        self.hand.add_card(card)
# getting the name of each player
    def get_name(self):
        return self.name
# getting the hand
    def get_hand(self):
        return self.hand
# returning the name
    def __str__(self):
        return self.name


# PlayerRank Class
class PlayerRank:
    def __init__(self, player, rank, suit):
        self.player = player
        self.rank = rank
        self.suit = suit
# returning the player
    def get_player(self):
        return self.player
# getting the rank
    def get_rank(self):
        return self.rank
# comparing each rank
    def __lt__(self, other):
        return self.rank < other.rank

    def get_suit(self):
        return self.suit
